parser/ParseIndexProperties.hpp

```diff
@@ -0,0 +1,267 @@
+                  const ParseKeyValue* invalid_property_node)
+  : is_index_description_valid_(is_index_description_valid),
+    reason_for_invalid_description_(reason_for_invalid_description),
+    invalid_property_node_(invalid_property_node),
+    index_sub_block_description_(nullptr) {
+  }
+
+  typedef std::unordered_map<std::string, ParseKeyValue::KeyValueType> valid_property_map_type;
+
+  bool hasValidIndexDescription() const {
+    return is_index_description_valid_;
+  }
+
+  const ParseKeyValue* getInvalidPropertyNode() const {
+    return invalid_property_node_;
+  }
+
+  std::string getReasonForInvalidIndexDescription() const {
+    return reason_for_invalid_description_;
+  }
+
+  const IndexSubBlockDescription* getIndexDescription() const {
+    return index_sub_block_description_.get();
+  }
+
+  void setIndexDescriptionAsInvalid(const std::string &reason, const ParseKeyValue *invalid_node = nullptr) {
+    is_index_description_valid_ = false;
+    index_sub_block_description_.reset();
+    reason_for_invalid_description_ = reason;
+    invalid_property_node_ = invalid_node;
+  }
+
+  virtual bool addCustomProperties(const PtrList<ParseKeyValue> *key_value_list) = 0;
+
+  bool hasDuplicateCustomProperty(const PtrList<ParseKeyValue> *key_value_list) {
+    // Define an existence map, which checks for duplicate properties in the key_value_list.
+    std::unordered_map<std::string, bool> is_duplicate_property;
+    for (const ParseKeyValue &key_value : *key_value_list) {
+      std::string key = ToLower(key_value.key()->value());
+      if (is_duplicate_property.find(key) != is_duplicate_property.end()) {
+        setIndexDescriptionAsInvalid(key + " is specified more than once", &key_value);
+        return true;  // duplicate found, return true.
+      } else {
+        is_duplicate_property[key] = true;
+      }
+    }
+    return false;  // no duplicates, return false.
+  }
+
+  bool hasInvalidCustomProperty(const PtrList<ParseKeyValue> *key_value_list) {
+    for (const ParseKeyValue &key_value : *key_value_list) {
+      std::string key = ToLower(key_value.key()->value());
+      if (valid_property_map_.find(key) == valid_property_map_.end()) {
+        setIndexDescriptionAsInvalid(key + " is an invalid property for this index", &key_value);
+        return true;  // this is not one of the valid keys, return true.
+      } else {
+        // Validate the type.
+        if (key_value.getKeyValueType() != valid_property_map_[key]) {
+          setIndexDescriptionAsInvalid(key + " property has an invalid type for this index", &key_value);
+          return true;  // this key has an unexpected type, return true.
+        }
+      }
+    }
+    return false;  // all valid keys, return false.
+  }
+
+ protected:
+  valid_property_map_type valid_property_map_;  // a map of index property constants and their associated type
+  bool is_index_description_valid_;
+  std::string reason_for_invalid_description_;
```

Comment:
The reason I didn't use enum was because this is supposed to be anyway a printable string and possible new indexes may define their own properties and corresponding validation and invalidation rules. A map would be additionally required if using an enum.